File: Assets/src/model/grasses/Mushroom.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using UnityEngine;

[System.Serializable]
[ObjectInfo(description: "Walk over it to Eat and reduce hunger by 5.")]
public class Mushroom : Grass, IActorEnterHandler {
  public static Item HomeItem => new ItemMushroom(4);
  public Mushroom(Vector2Int pos) : base(pos) {
  }

  public void HandleActorEnter(Actor actor) {
    var player = GameModel.main.player;
    if (actor == player) {
      player.hunger -= 5;
      Kill(player);
    }
  }

  public static bool CanOccupy(Tile tile) {
    var floor = tile.floor;
    // hugging at least one 4-neighbor wall
    var cardinalNeighbors = floor.GetCardinalNeighbors(tile.pos);
    var isHuggingWall = cardinalNeighbors.Any((pos) => pos is Wall);
    var isGround = tile is Ground;
    var isNotOccupied = tile.grass == null;
    
    return isHuggingWall && isGround && isNotOccupied;
  }
}

File: web/src/model/grasses/Mushroom.ts
import { Grass } from './Grass';
import { ACTOR_ENTER_HANDLER, type IActorEnterHandler } from '../../core/types';
import { Ground, Wall } from '../Tile';
import { GameModelRef } from '../GameModelRef';
import { ItemMushroom } from '../items/ItemMushroom';
import { ItemOnGround } from '../ItemOnGround';
import { entityRegistry } from '../../generator/entityRegistry';
import type { Vector2Int } from '../../core/Vector2Int';
import type { Tile } from '../Tile';

/**
 * Walk over it to harvest. Player picks up an ItemMushroom.
 * Port of C# Mushroom.cs.
 */
export class Mushroom extends Grass implements IActorEnterHandler {
  readonly [ACTOR_ENTER_HANDLER] = true as const;

  constructor(pos: Vector2Int) {
    super(pos);
  }

  static canOccupy(tile: Tile): boolean {
    const floor = tile.floor;
    if (!floor) return false;
    const isHuggingWall = floor.getCardinalNeighbors(tile.pos).some(t => t instanceof Wall);
    return isHuggingWall && tile instanceof Ground && tile.grass == null;
  }

  handleActorEnter(actor: any): void {
    const player = GameModelRef.main.player;
    if (actor === player) {
      this.becomeItemInInventory(new ItemMushroom(1));
    }
  }

  /** Kill self and add item to player inventory (or drop on ground). */
  private becomeItemInInventory(item: import('../Item').Item): void {
    const player = GameModelRef.main.player;
    const floor = this.floor;
    this.kill(player);
    if (!player.inventory.addItem(item, this)) {
      if (floor) {
        floor.put(new ItemOnGround(this.pos, item, this.pos));
      }
    }
  }
}

entityRegistry.register('Mushroom', Mushroom);
